### PR TITLE
feat: cicd pipeline bind queue

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
cicd pipeline bind queue

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?id=327948&iterationID=1354&pId=0&type=REQUIREMENT)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  Support cicd pipeline bind queue            |
| 🇨🇳 中文    |      cicd流水线绑定队列        |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by littlejiancc — APPROVED** [review]

File: internal/apps/dop/services/pipeline/pipeline.go [hunk 1]
```diff
@@ -396,11 +412,19 @@ func (p *Pipeline) ConvertPipelineToV2(pv1 *apistructs.PipelineCreateRequest) (*
 	pv2.NormalLabels = normalLabels
 
 	// clusterName
-	pj, err := p.bdl.GetProject(app.ProjectID)
-	if err != nil {
+	pj, apiErr := p.projectSvc.Get(context.Background(), app.ProjectID)
+	if apiErr != nil {
 		return nil, apierrors.ErrGetProject.InternalError(err)
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply internal/apps/dop/services/pipeline/pipeline.go]: adjusted

**Comment by chengjoey** [comment]
#https://github.com/erda-project/docs/pull/246

File: internal/apps/dop/providers/queue/interface.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	queue, err := p.QueueManager.CreateQueue(context.Background(), &pb.QueueCreateRequest{
+		Name:             p.makeProjectLevelQueueName(workspace, project.Name),
+		ClusterName:      resourceConfig.ClusterName,
+		PipelineSource:   apistructs.PipelineSourceDice.String(),
+		ScheduleStrategy: apistructs.ScheduleStrategyInsidePipelineQueueOfFIFO.String(),
+		MaxCPU:           queueResource.MaxCPU,
+		MaxMemoryMB:      queueResource.MaxMemoryMB,
+		Concurrency:      queueResource.Concurrency,
+		Labels: map[string]string{
+			"projectName": project.Name,
```
Comment (sfwn) [inline internal/apps/dop/providers/queue/interface.go:52]: also add project id

Comment (sfwn) [reply internal/apps/dop/providers/queue/interface.go:52]: use variable to define keys.

File: internal/apps/dop/providers/queue/calculate.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/erda-project/erda/apistructs"
+)
+
+type ProjectQueueResource struct {
+	Concurrency int64
+	MaxCPU      float64
+	MaxMemoryMB float64
+}
+
+func maxInt64(a, b int64) int64 {
```
Comment (sfwn) [inline internal/apps/dop/providers/queue/calculate.go:31]: use pkg.numeral

File: internal/apps/dop/providers/queue/calculate.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	apps, err := p.bdl.CountAppByProID(project.ID)
+	if err != nil {
+		return 0, err
+	}
+	concurrency := apps * 2
+	return maxInt64(concurrency, p.Cfg.DefaultQueueConcurrency), nil
+}
+
+func (p *provider) calculateMaxCpu(resourceConfig *apistructs.ResourceConfigInfo) (float64, error) {
+	return maxFloat64(p.Cfg.DefaultQueueCPU, resourceConfig.CPUQuota), nil
```
Comment (sfwn) [inline internal/apps/dop/providers/queue/calculate.go:57]: I think you should compare with 0 if no resource available.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/providers/queue/provider.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/erda-project/erda-infra/base/servicehub"
+	"github.com/erda-project/erda-infra/pkg/transport"
+	"github.com/erda-project/erda-proto-go/core/pipeline/queue/pb"
+	"github.com/erda-project/erda/bundle"
+)
+
+type config struct {
+	DefaultQueueConcurrency int64   `env:"DEFAULT_QUEUE_CONCURRENCY" default:"20"`
+	DefaultQueueCPU         float64 `env:"DEFAULT_QUEUE_CPU" default:"30"`
+	DefaultQueueMemoryMB    float64 `env:"DEFAULT_QUEUE_MEMORY_MB" default:"30000"` // default: 30GB
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply internal/apps/dop/providers/queue/provider.go:31]: adjusted

File: internal/apps/dop/providers/queue/interface.go (hunk 2 above)
Comment (chengjoey) [reply internal/apps/dop/providers/queue/interface.go:52]: adjusted

File: internal/apps/dop/providers/queue/calculate.go (hunk 3 above)
Comment (chengjoey) [reply internal/apps/dop/providers/queue/calculate.go:31]: adjusted

File: internal/apps/dop/providers/queue/calculate.go (hunk 4 above)
Comment (chengjoey) [reply internal/apps/dop/providers/queue/calculate.go:57]: compared

**Review by sfwn — APPROVED** [review]

**Pull request merged**